docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,251 @@
+
+Pros:
+- The deletion can be applied earlier than the previous options.
+    - Only applies if the deletion request cancel period is less than the last time interval for compaction is.
+Cons:
+- Added coupling between the compaction and the DeletedSeriesCleaner.
+- Might block compaction for a short time while doing the deletion.
+
+
+
+Once all the applicable blocks have been rewritten without the deleted data, the deletion request state moves to the `Processed` state. Once in this state, the queriers will still have to perform query time filtering using the tombstones until the old blocks that were marked for deletion are no longer queried by the queriers. This will mean that the query time filtering will last for an additional length of `-compactor.deletion-delay + -compactor.cleanup-interval +  -blocks-storage.bucket-store.sync-interval` in the `Processed` state. Once that time period has passed, the queriers should no longer be querying any of the old blocks that were marked for deletion. The tombstone will no longer be used after this.
+
+
+#### Handling failed/unfinished delete jobs:
+
+Deletions will be completed and the tombstones will be deleted only when the DeletedSeriesCleaner iterates over all blocks that match the time interval and confirms that they have been re-written without the deleted data.  Otherwise, it will keep iterating over the blocks and process the blocks that haven't been rewritten according to the information in the `meta.json` file. In case of any failure that causes the deletion to stop, any unfinished deletions will be resumed once the service is restarted. If the block rewrite was not completed on a particular block, then the original block will not be marked for deletion. The compactor will continue to iterate over the blocks and process the block again.
+
+
+#### Tenant Deletion API
+
+If a request is made to delete a tenant, then all the tombstones will be deleted for that user.
+
+
+
+## Current Open Questions:
+
+- If the start and end time is very far apart, it might result in a lot of the data being re-written. Since we create a new block without the deleted data and mark the old one for deletion, there may be a period of time with lots of extra blocks and space used for large deletion queries.
+- There will be a delay between the deletion request and the deleted data being filtered during queries.
+    - In Prometheus, there is no delay.
+    - One way to filter out immediately is to load the tombstones during query time but this will cause a negative performance impact.
+- Adding limits to the API such as:
+    - The number of deletion requests per day,
+    - Number of requests allowed at a time
+    - How wide apart the start and end time can be.
```

Comment:
Not sure about this. I think should be possible to request deletion for all the customer's data (why shouldn't?).

Reply:
I guess it doesn't make too much sense. Removed it.